### PR TITLE
Changed pivot from a template argument to a function argument

### PR DESCRIPTION
**Pull request by tfalders**
Addresses this comment: https://github.com/ROCmSoftwarePlatform/rocSOLVER/pull/305#discussion_r687059768

Feel free to merge when ready.

### CODE REVIEW

File: library/src/lapack/roclapack_getf2_optimized.cpp [hunk 1]
```diff
@@ -40,7 +39,7 @@ ROCSOLVER_KERNEL void __launch_bounds__(GETF2_MAX_THDS)
     // batch instance
     T* A = load_ptr_batch<T>(AA, id, shiftA, strideA);
     rocblas_int* ipiv;
-    if(pivot)
+    if(ipivA != nullptr)
```
Comment (jzuniga-amd) [inline library/src/lapack/roclapack_getf2_optimized.cpp:42]: what is the difference between using the argument `pivot` or transforming nullptr into a boolean?
Are you thinking the code could be somehow better if we use one less parameter?

File: library/src/lapack/roclapack_getrf.hpp [hunk 2]
```diff
@@ -437,38 +424,53 @@ rocblas_status rocsolver_getrf_template(rocblas_handle handle,
...
+                hipLaunchKernelGGL(getrf_check_singularity<T>, grid, threads, 0, stream, n, j + k,
+                                   jb1, A, shiftA, lda, strideA, ipiv, shiftP, strideP, iipiv,
+                                   iinfo, info);
+            }
+            else
+            {
+                // factorize inner block panel
+                rocsolver_getf2_template<ISBATCHED, T>(
+                    handle, m - j - k, jb1, A, shiftA + idx2D(j + k, j + k, lda), lda, strideA,
+                    nullptr, 0, jb1, iinfo, batch_count, scalars, pivotval, pivotidx);
```
Comment (jzuniga-amd) [inline library/src/lapack/roclapack_getrf.hpp:468]: is not `iipiv = nullptr` a necessary and sufficient condition for `getf2_template` to avoid pivoting? I think you can get this call out of the if statement. 

(also, if you pass `pivot` to the templates, you can have the call out of the if statement regardless of the value of `iipiv`. After all, `pivot` should be set to the right value right at the beginning: at the C API wrapper level).

File: library/src/lapack/roclapack_getf2.cpp [hunk 3]
```diff
@@ -127,7 +128,7 @@ rocblas_status rocsolver_sgetf2_npvt(rocblas_handle handle,
                                      rocblas_int* info)
 {
     rocblas_int* ipiv = nullptr;
-    return rocsolver_getf2_impl<false, float>(handle, m, n, A, lda, ipiv, info);
+    return rocsolver_getf2_impl<float>(handle, m, n, A, lda, ipiv, info, false);
```
[the remark this answers is not in the document]
Comment (tfalders) [reply library/src/lapack/roclapack_getf2.cpp:131]: There were three main reasons for why I left out the `pivot` argument in the template functions, which I arrived at during the work on getri_npvt:
1. I have an unverified belief that using fewer arguments makes kernel launches faster, and I extended that approach to host code.
2. Because of argument checking, I was fairly confident that `ipiv` should only be a `nullptr` in the npvt case. At the time, I hadn't thought of workspace arrays like `iipiv`, which throws a wrench in that idea.
3. I didn't want to add a `pivot` argument to `getrs`, which handles the pivoting for `getri_outofplace`. For consistency, I didn't add it to other functions either.

The first two reasons are pretty weak, so I'm happy to add the argument in. I guess the only question is if we're okay adding it to getrs as well, even though it does not currently support npvt routines.

[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/lapack/roclapack_getf2.cpp:131]: I see...

I think it is OK. `getrs` actually supports npvt cases!! it just does not need to have explicit npvt APIs; it behaves according to how A was factorized: If A=PLU, then `getrs` apply the permutations, if A=LU then `getrs` does not care about permutations... 
The thing here is how we let `getrs` know that it needs to apply permutations? Right now we use
`pivot = (ipiv != nullptr)` inside `getrs_template`, but we can also pass the already set argument `pivot` to `getrs_template`...

As mentioned before, we know that both approaches will work (they are equivalent as the arg checking won't let us have `pivot = true && ipiv = nullptr`, for example), I just think that we should use only one approach and not combine them. For the reasons I exposed before (and in the `getri_npvt` PR) I feel more comfortable with cascading down the argument `pivot` everywhere it is needed... is like the `shift`, `stride` and `batch_count` arguments, you just set them once in the implementation of the C API, and pass them everywhere so you don't have to take care of them anymore. 

So, in conclusion... I would be happy with either option, combining `pivot` and the `ipiv != nullptr` or only using `pivot`. I leave it up to you :)

Let me know when you are ready for me to merge this PR.

[the remark this answers is not in the document]
Comment (tfalders) [reply library/src/lapack/roclapack_getf2.cpp:131]: I've made the change. Everything is compiling and running correctly (I haven't tested the optimized kernels, though, since they take such a long time to compile).

Looks like merging develop into my branch has caused the diff to include all the changes made in the getri_npvt PR. I'd recommend isolating the "Add pivot argument to template functions" commit for review since that's where my actual changes were made.

Ready to merge when you are.

**Comment by jzuniga-amd** [comment]
Looks good!!
I will merge this now... thanks Troy.
